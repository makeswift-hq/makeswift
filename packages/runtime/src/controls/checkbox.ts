export type CheckboxControlData = boolean

export const CheckboxControlType = 'makeswift::controls::checkbox'

type CheckboxControlConfig = {
  label?: string
  defaultValue?: boolean
}

export type CheckboxControlDefinition<C extends CheckboxControlConfig = CheckboxControlConfig> = {
  type: typeof CheckboxControlType
  config: C
}

export function Checkbox<C extends CheckboxControlConfig>(
  config: C = {} as C,
): CheckboxControlDefinition<C> {
  return { type: CheckboxControlType, config }
}
