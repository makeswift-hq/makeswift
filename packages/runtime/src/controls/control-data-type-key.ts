export const ControlDataTypeKey = '@@makeswift/type'
