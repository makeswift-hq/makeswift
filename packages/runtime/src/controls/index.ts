export * from './color'
export * from './combobox'
export * from './control'
export * from './icon-radio-group'
export * from './image'
export * from './link'
export * from './list'
export * from './select'
export * from './shape'
export * from './slot'
export * from './style'
export * from './style-v2'
export * from './rich-text'
export * from './rich-text-v2'
export * from './typography'
