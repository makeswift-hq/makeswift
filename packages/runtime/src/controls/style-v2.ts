import { BoxModel } from 'css-box-model'
import { ControlDefinition, ControlDefinitionData } from './control'
import { PropController } from '../prop-controllers/base'
import { Send } from '../prop-controllers/instances'
import { ResponsiveValue } from './types'
import { CSSObject } from '@emotion/serialize'
import { useStyle } from '../runtimes/react/use-style'

export const StyleV2ControlType = 'makeswift::controls::style-v2'

export type StyleV2CSSObject = CSSObject
export const unstable_useStyleV2ClassName = useStyle

type StyleV2ControlConfig<T extends ControlDefinition = ControlDefinition> = {
  type: T
  getStyle(item: ControlDefinitionData<T> | undefined): CSSObject
}

export type StyleV2ControlDefinition = {
  type: typeof StyleV2ControlType
  config: StyleV2ControlConfig
}

export function unstable_StyleV2<C extends StyleV2ControlConfig>(
  config: C,
): StyleV2ControlDefinition {
  return { type: StyleV2ControlType, config }
}

/**
 * TODO: Rewrite `ControlDefinitionData` as a tail recursive type.
 *
 * Explanation: Making `StyleV2ControlData` generic causes us to hit a limitation of TS conditional types for the definition of `ControlDefinitionData`.
 * We get "Type instantiation is excessively deep and possibly infinite."
 *
 * Helpful links:
 * * The PR changing the limit of conditional types: https://github.com/microsoft/TypeScript/pull/45711
 * * TS 4.5 docs releasing this limitation: https://www.typescriptlang.org/docs/handbook/release-notes/typescript-4-5.html#tail-recursion-elimination-on-conditional-types
 * * A SO user implementing a tail recursive type: https://stackoverflow.com/questions/72370456/tail-recursion-elimination-on-conditional-types-doesnt-work
 *
 *
 * This is what the type should be:
 * ```
 * export type StyleV2ControlData<T extends StyleV2ControlDefinition = StyleV2ControlDefinition> =
 *   ResponsiveValue<ControlDefinitionData<T['config']['type']>>
 * ```
 */
export type StyleV2ControlData = ResponsiveValue<any>

export const StyleV2ControlMessageType = {
  CHANGE_BOX_MODEL: 'makeswift::controls::style::message::change-box-model',
} as const

type StyleV2ControlItemBoxModelChangeMessage = {
  type: typeof StyleV2ControlMessageType.CHANGE_BOX_MODEL
  payload: { boxModel: BoxModel | null }
}

export type StyleV2ControlMessage = StyleV2ControlItemBoxModelChangeMessage

export class StyleV2Control extends PropController<StyleV2ControlMessage> {
  constructor(send: Send<StyleV2ControlMessage>) {
    super(send)
  }

  changeBoxModel(boxModel: BoxModel | null): void {
    this.send({ type: StyleV2ControlMessageType.CHANGE_BOX_MODEL, payload: { boxModel } })
  }

  recv() {}
}
