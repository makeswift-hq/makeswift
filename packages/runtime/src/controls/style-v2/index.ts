export * from './style-v2'
