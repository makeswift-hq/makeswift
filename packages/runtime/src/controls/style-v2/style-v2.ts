import {
  ControlDefinition,
  SerializationSchema,
  StyleV2Definition as BaseStyleV2Definition,
  StyleV2PropDefinition,
  StyleV2Control,
  type StyleV2Config,
  type DeserializedRecord,
} from '@makeswift/controls'

import { CSSObject } from '@emotion/serialize'

type PropDefinition = StyleV2PropDefinition
type Config<Prop extends PropDefinition> = StyleV2Config<Prop, CSSObject>

export class StyleV2Definition<
  Prop extends PropDefinition = PropDefinition,
> extends BaseStyleV2Definition<Prop, CSSObject> {
  static deserialize(
    data: DeserializedRecord,
    deserializeCallback: (r: DeserializedRecord) => ControlDefinition,
  ): StyleV2Definition {
    if (data.type !== StyleV2Definition.type) {
      throw new Error(`StyleV2: expected type ${StyleV2Definition.type}, got ${data.type}`)
    }

    const {
      config: { type, ...config },
    } = StyleV2Definition.schema({
      typeDef: SerializationSchema.deserializedRecord,
    }).definition.parse(data)

    const typeDef = deserializeCallback(type)

    return unstable_StyleV2({ type: typeDef, ...config })
  }
}

export function unstable_StyleV2<Prop extends PropDefinition>(
  config: Config<Prop>,
): StyleV2Definition<Prop> {
  return new StyleV2Definition(config)
}

export { StyleV2Control }
