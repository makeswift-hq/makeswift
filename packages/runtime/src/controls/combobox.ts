import { Data } from '@makeswift/controls'

export const ComboboxControlType = 'makeswift::controls::combobox'

type ComboboxControlOption<T extends Data> = { id: string; label: string; value: T }

type ComboboxControlConfig<T extends Data = Data> = {
  label?: string
  getOptions(query: string): ComboboxControlOption<T>[] | Promise<ComboboxControlOption<T>[]>
}

export type ComboboxControlDefinition<C extends ComboboxControlConfig = ComboboxControlConfig> = {
  type: typeof ComboboxControlType
  config: C
}

export type ComboboxControlDefinitionOption<T extends ComboboxControlDefinition> =
  T['config'] extends ComboboxControlConfig<infer U> ? U : never

export type ComboboxControlData<T extends ComboboxControlDefinition = ComboboxControlDefinition> =
  ComboboxControlOption<ComboboxControlDefinitionOption<T>>

export function Combobox<T extends Data, C extends ComboboxControlConfig<T>>(
  config: C & {
    getOptions(query: string): ComboboxControlOption<T>[] | Promise<ComboboxControlOption<T>[]>
  },
): ComboboxControlDefinition<C> {
  return { type: ComboboxControlType, config }
}
