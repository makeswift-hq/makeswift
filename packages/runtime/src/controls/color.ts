import { match } from 'ts-pattern'
import type { CopyContext } from '../state/react-page'
import type { ColorData } from './types'
import { ControlDataTypeKey } from './control-data-type-key'

export const ColorControlDataTypeKey = ControlDataTypeKey

export const ColorControlDataTypeValueV1 = 'color::v1'

export type ColorControlDataV0 = ColorData

export type ColorControlDataV1 = ColorControlDataV0 & {
  [ColorControlDataTypeKey]: typeof ColorControlDataTypeValueV1
}

export type ColorControlData = ColorControlDataV0 | ColorControlDataV1

export const ColorControlType = 'makeswift::controls::color'

type ColorControlConfig = {
  label?: string
  labelOrientation?: 'horizontal' | 'vertical'
  defaultValue?: string
  hideAlpha?: boolean
}

export type ColorControlDefinition<C extends ColorControlConfig = ColorControlConfig> = {
  type: typeof ColorControlType
  config: C
  version?: 1
}

export function Color<C extends ColorControlConfig>(config = {} as C): ColorControlDefinition<C> {
  return { type: ColorControlType, config, version: 1 }
}

export function copyColorData(
  value: ColorControlData | undefined,
  context: CopyContext,
): ColorControlData | undefined {
  if (value == null) return value

  return match(value)
    .with({ [ColorControlDataTypeKey]: ColorControlDataTypeValueV1 }, val => ({
      ...val,
      swatchId: context.replacementContext.swatchIds.get(val.swatchId) ?? val.swatchId,
    }))
    .otherwise(val => ({
      ...val,
      swatchId: context.replacementContext.swatchIds.get(val.swatchId) ?? val.swatchId,
    }))
}
