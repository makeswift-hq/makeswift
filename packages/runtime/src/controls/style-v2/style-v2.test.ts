import { unstable_IconRadioGroup, deserializeRecord } from '@makeswift/controls'

import { unstable_StyleV2, StyleV2Definition } from './style-v2'
import { deserializeUnifiedControlDef } from '../../builder'

describe('StyleV2', () => {
  test('serialization', () => {
    const definition = unstable_StyleV2({
      type: unstable_IconRadioGroup({
        label: 'Alignment',
        options: [
          {
            icon: unstable_IconRadioGroup.Icon.TextAlignLeft,
            label: 'Left Align',
            value: 'left',
          },
          {
            icon: unstable_IconRadioGroup.Icon.TextAlignCenter,
            label: 'Center Align',
            value: 'center',
          },
          {
            icon: unstable_IconRadioGroup.Icon.TextAlignRight,
            label: 'Right Align',
            value: 'right',
          },
          {
            icon: unstable_IconRadioGroup.Icon.TextAlignJustify,
            label: 'Justify',
            value: 'justify',
          },
        ],
        defaultValue: 'left',
      }),
      getStyle(textAlign) {
        return { textAlign }
      },
    })

    const [serialized, transferables] = definition.serialize()
    transferables.forEach((port: any) => port.close())

    const deserialized = StyleV2Definition.deserialize(
      deserializeRecord(serialized),
      deserializeUnifiedControlDef,
    )

    expect(deserialized).toMatchSnapshot()
  })
})
