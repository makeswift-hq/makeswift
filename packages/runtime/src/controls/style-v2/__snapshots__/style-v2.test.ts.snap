// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StyleV2 serialization 1`] = `
StyleV2Definition {
  "config": {
    "getStyle": [Function],
    "type": IconRadioGroupDefinition {
      "config": {
        "defaultValue": "left",
        "label": "Alignment",
        "options": [
          {
            "icon": "TextAlignLeft",
            "label": "Left Align",
            "value": "left",
          },
          {
            "icon": "TextAlignCenter",
            "label": "Center Align",
            "value": "center",
          },
          {
            "icon": "TextAlignRight",
            "label": "Right Align",
            "value": "right",
          },
          {
            "icon": "TextAlignJustify",
            "label": "Justify",
            "value": "justify",
          },
        ],
      },
    },
  },
}
`;
